L82-core-event-engine.md

```diff
@@ -0,0 +1,478 @@
+gRPC Core EventEngine API
+----
+* Author: AJ Heller (@drfloob)
+* Approver: Mark Roth (@markdroth)
+* Status: In Review
+* Implemented in: gRPC Core
+* Last updated: 2021.06.15
+* Discussion at: https://groups.google.com/g/grpc-io/c/EamJ_ae_db0
+
+## Abstract
+
+This work replaces gRPC Core's iomgr with a public interface for custom,
+pluggable implementations which we're calling EventEngines. EventEngines are
+tasked with providing all cross-platform I/O, task execution, and DNS resolution
+functionality for gRPC Core and its wrapped languages.
+
+Unlike the current polling model where threads are borrowed from the
+application, EventEngines will bring their own polling and callback execution
+threads. This public API will make it easier to integrate gRPC into external
+event loops, it will support siloing events between gRPC channels and servers,
+and it will provide another way to support the C++ Callback API.
+
+## Background
+
+gRPC Core and its wrapped languages rely internally on an event management
+framework called iomgr. A small collection of homegrown, platform-specific
+implementations exist, allowing I/O functionality to be customized to the
+various platforms that gRPC supports (IOCP is used on Windows, epoll on Linux,
+poll on Mac, etc).
+
+Over the past few years, there have been multiple projects aimed at improving or
+replacing iomgr. The EventEngine effort takes a slightly different approach to
+things, but otherwise shares the same rationale as the previous effort.
+
+### Related Proposals:
+
+* This work supersedes [L69: EventManger API](https://github.com/grpc/proposal/pull/182)
+* Supports [L67: C++ callback-based asynchronous API](https://github.com/grpc/proposal/pull/180) as a more performant solution than provided by the [CallbackAlternativeCQ](https://github.com/grpc/grpc/pull/25169).
+
+## Proposal
+
+### Concepts
+
+An EventEngine is roughly composed of:
+
+* A task runner which asynchronously executes callables with optional
+  scheduling.
+* An asynchronous DNS resolver.
+* Network listener logic that accepts incoming connection requests, and
+  performs network-server-related I/O operations.
+* Client logic, support for making outgoing network connections.
+
+### Common Data Types
+
+```cpp
+namespace grpc_event_engine {
+namespace experimental {
+class EventEngine {
+ public:
+  /// A basic callable function. The first argument to all callbacks is an
+  /// absl::Status indicating the status of the operation associated with this
+  /// callback. Each EventEngine method that takes a callback parameter, defines
+  /// the expected sets and meanings of statuses for that use case.
+  using Callback = std::function<void(absl::Status)>;
+  /// A callback handle, used to cancel a callback.
+  struct TaskHandle {
```

Comment:
This likely needs to have some more space allocated to it.

Right now an implementation atop the internal EventManager2 would be forced to allocate per timer, which is probably unacceptable.

Suggest a handle size of at least 2 intptr_t's.

Reply:
Pinging @soheilhy to verify my understanding

Reply:
+1 please allow another QWORD in addition to the key so that we can use it for a more efficient implementation internally.

Reply:
I've changed the definition to `intptr_t keys[2]`.